web/massastation/src/components/SidePanels/Upload/Upload.tsx

```diff
@@ -7,8 +8,132 @@ import {
+}
 
 export default function Upload() {
+  const form = useRef(null);
+  const [formError, setFormError] = useState<IFormError | null>(null);
+  const [fileError, setFileError] = useState<string | null>(null);
+  const [accountsError, setAccountError] = useState<string | null>(null);
+  const [file, setFile] = useState<File | null>(null);
+  const nickname = useAccountStore((state) => state.nickname);
+
+  const { mutate: mutableUpload, isLoading: uploadLoading } = usePut<
+    FormData,
+    IUploadResponse
+  >('websiteUploader/prepare', {
+    'Content-Type': 'multipart/form-data',
+  });
+
+  async function validate(e: SyntheticEvent): Promise<boolean> {
+    const formObject = parseForm<IFormObject>(e);
+    const { websiteName, description } = formObject;
+
+    // reset errors
+    setFileError(null);
+    setFormError(null);
+    setAccountError(null);
+
+    if (nickname === null) {
+      setAccountError(Intl.t('search.errors.no-nickname'));
+      return false;
+    }
+
+    if (!websiteName) {
+      setFormError({ websiteName: Intl.t('search.errors.no-website-name') });
+      return false;
+    }
+
+    if (!validateWebsiteName(websiteName)) {
+      setFormError({
+        websiteName: Intl.t('search.errors.invalid-website-name'),
+      });
+      return false;
+    }
+
+    if (description === '') {
+      setFormError({ description: Intl.t('search.errors.no-description') });
+      return false;
+    }
+
+    if (!validateDescriptionLength(description)) {
+      setFormError({
+        description: Intl.t('search.errors.description-too-long'),
+      });
+      return false;
+    }
+
+    if (!validateWebsiteDescription(description)) {
+      setFormError({
+        description: Intl.t('search.errors.invalid-description'),
+      });
+      return false;
+    }
+
+    if (!file) {
+      setFileError(Intl.t('search.errors.no-file'));
+      return false;
+    }
+
+    if (!validateFileExtension(file?.name)) {
+      setFileError(Intl.t('search.errors.invalid-file-extension'));
+      return false;
+    }
+
+    if (!(await validateFileContent(file))) {
+      setFileError(Intl.t('search.errors.invalid-file-content'));
+      return false;
+    }
+
+    return true;
+  }
+
+  function uploadWebsite(e: SyntheticEvent) {
+    const { websiteName, description } = parseForm<IFormObject>(e);
+    const bodyFormData = new FormData();
+    bodyFormData.append('url', websiteName);
+    bodyFormData.append('description', description); // Add the website description to the form data
+    bodyFormData.append('nickname', nickname as string); // we force compiler type because we know it's not null
+    bodyFormData.append('zipfile', file as File); // we force compiler type of `file` because we know it's not null
+    mutableUpload(bodyFormData);
+  }
+
+  function handleSubmit(e: SyntheticEvent) {
+    e.preventDefault();
+
+    validate(e).then((isValid) => {
+      if (isValid) {
+        uploadWebsite(e);
+      }
+    });
+  }
```

Comment:
Can we translate in async/await instead of then... 🤔 

```suggestion
  async function handleSubmit(e: SyntheticEvent) {
    e.preventDefault();

    let isValid = await validate(e)
    if (!isValid) return;
    uploadWebsite(e);
   }
```

Reply:
no we can't because handleSubmit is in a onSubmit and it must be sync